README
Well, this is Pycorn.

Pycorn is an interpreted operating system written in Python, intended to
recreate the style of learning available on 16-bit microcomputers. Device
drivers, file systems, network protocols can all be implemented in Python with
no C or assembler code.

The goal of the project is to create a usable interpreter-centric OS
environment with the capabilities of modern computers; allowing a smooth
learning curve from the acquisition of basic programming skills right up to
implementing new OS-level functionality and drivers. Performance is not a goal,
though it would be nice :) Current development is targeted at a variety of ARM
platforms due to the relative simplicity of the hardware.

LICENSE
=======

---------------------------------------------------------------------
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
---------------------------------------------------------------------

The directory 'makepp' is a verbatim copy of the original makepp source,
distributed here for convenience. makepp has its own licensing terms in
makepp/LICENSE.

HOW IT ALL FITS TOGETHER
========================

Pycorn uses a cross-compiling gcc and the newlib embedded C library.  This
provides enough 'stuff' to compile the main python interpreter into a library.

To actually make a bootable image a linker script and entry point are required
which are implemented by pycorn. These are architecture specific and include a
few machine specific details: a small linker script with some platform physical
addresses includes the main architecture linker script. The entry point uses a
small amount of assembly then switches to position-independent C code for the
majority of platform setup, then calls a generic C library setup function
written in C.

Right now we only have them for Marvell's PXA270 system-on-chip, as used in the
Gumstix Verdex development board (http://www.gumstix.com). This board can be
simulated by QEMU. A main() function is also required which invokes the python
interpreter with some arguments, and though this is written in C it is
portable.

Of course, it won't actually do anything visible; newlib has no actual IO
facilities. A platform-specific serial debug driver is used for now, hooked up
to newlib's stdin/stdout/stderr.

And.. that's it for now. There is no way to import any modules which aren't
built into the interpreter as all file access fails right now, and Python's
builtin readline is really stupid so you can't even backspace.. but it works!

BINARY RELEASES
===============

Pycorn is not very useful as a binary image at this point, because the builtin
Python code is not yet finished, and so it is expected that people will want to
play around with the Python code and build a new image from it.

However, the actual native binary contains only the minimum amount of code to
start up the interpreter and begin to import stuff, and binary releases contain
a prebuilt copy of this binary. This means you don't need a compiler.

REQUIREMENTS
============

Pycorn requires several things in order to build, even if you are using a
binary release:

1) Add the makepp directory in the pycorn source tree to your PATH. makepp is a
nice make replacement with lots of nifty features; used primarily because I
hate make's handling of subdirectories. See http://makepp.sourceforge.net/ for
their project page. I've included a version in the tree because it's a CVS
snapshot, and an uncommon tool.

2) If you aren't using a binary release, you need an arm-eabi crosscompiler
toolchain, and newlib 1.17. To build one automatically, install the header
files for libgmp, libmpfr and ncurses (the dev packages for them from your
distro), then run:

cd toolchain; makepp TOOLSPREFIX=/dir/to/install/to

Once it's built, add the bin directory inside TOOLSPREFIX to your path. If you
install one from somewhere else, newlib must be built to expect reentrant
syscalls and without its own embedded syscalls or it will not work; better to
use this one.

3) If you aren't using a binary release, install autoconf. Python's configure
script needs regeneration after patching for cross-compilability.

BUILDING
========

If you are using a binary release, skip step 1.

1) Run "makepp seeds/hello/hello.uimage". This is a "hello world" program which
should load into u-boot and print on the default platform serial port.

1a) If you have an actual Gumstix Verdex board you should be able to load this
by serial, tftp, or MMC, and execute it with bootm.

1b) If you don't, you can simulate it under QEMU by doing
"makepp seeds/hello/run", assuming you have QEMU installed. The bootloader will
 be configured automatically so it should just run. Hit C-a x (C-a a x under
screen) to kill QEMU.

2) Run "makepp seeds/pycorn/pycorn.uimage". This is the python interpreter,
with the minimum amount of code to boot pycorn built into it.

3a) Again, if you have an actual Verdex board it should run from u-boot.

3b) Otherwise, "makepp seeds/pycorn/run" to launch it in QEMU.


makepp/ChangeLog
2010-11-17  Daniel Pfeiffer  <occitan@esperanto.org>

Snapshot makepp-1.50.101117.tgz, version 1.50-101117:18-1018:5-0929:8 on CPAN.

	* *: Cleanly distinguish between lexing, parsing and scanning.
	Fix skip-word without recursion and provide new variant for
	(sub)shell.

	* pod/makepp_compatibility.pod: Add 5.12.2 and more CPAN-tester
	results.

2010-10-18  Daniel Pfeiffer  <occitan@esperanto.org>

	* *: Rename ActionParser to Lexer and only refer to it by that
	name, to get rid of previous chaos.

	* Mpp/ActionParser/Specific.pm:
	* Mpp/ActionParser/Legacy.pm: Deprecate useless subclasses that
	make it hard to get skip-word right.

	* Mpp/Rule.pm (scan_action): Remove unused func.
	(find_all_targets_dependencies): Fix missing return value (thx to
	thebamaman).

2010-09-29  Daniel Pfeiffer  <occitan@esperanto.org>

	* *: @Mpp::Text::N replaces individual numeric constants.

	* Mpp/FileOpt.pm (load_build_info_file):
	* makeppinfo: With -f, --force show outdated info.

2010-09-18  Daniel Pfeiffer  <occitan@esperanto.org>

	* Mpp/File.pm (path_file_info): On Win translate \\ownhost\c$ to
	c: to find same file under different names and reduce $ problems.

	* Mpp/Cmds.pm (c_cp): Respect $MAKEPP_LN_CP.
	(c_mkdir): Handle --mode like Unix variant.

2010-09-13  Daniel Pfeiffer  <occitan@esperanto.org>

	* Mpp/Text.pm (@common_opts): Also handle --help, getting it from
	__DATA__ everywhere.

	* Mpp.pm (perform): Don't duplicate error message at end.

	* Mpp/Makefile.pm (skip_makefile_until_else_or_endif): Statements
	like sub must end with space -- don't be fooled by a rule for
	sub-target.

	* makepp: Don't silently ignore gmake opts.  Instead
	$MAKEPP_IGNORE_OPTS allows controlling that, even -R is possible.

2010-09-03  Daniel Pfeiffer  <occitan@esperanto.org>

	* makepp (build_dependencies_done): Don't remove a file that
	wasn't there when we last checked, warn if it appeared.

	* Mpp/Recursive.pm:
	* recursive_makepp: Fix protocol error on -j failure.

	* Mpp/Subs.pm (f_origin): Same order as expansion and 'global'.

	* Mpp/Makefile.pm (MAKEPP_VERSION): New variable.
	($if_re): Refactor repeated regexp.

	*: Move version (and beta counting, now same for all progs and
	like snapshot names) to Mpp::Text.  Short opt -V.

2010-08-22  Daniel Pfeiffer  <occitan@esperanto.org>

	* Mpp/ActionParser.pm (parse_rule): Handle Shell keywords
	directly, so we catch variable assignment before the command.

	* Mpp/Scanner/C.pm (xscan_file): Have an informative warning about
	ignoring #include MACRO.

	* Many small compatibility changes:
	Provide functions abspath, realpath, and, or.
	Function suffix returns the dot too.
	Always set MAKEFLAGS.
	Allow special variables to be empty outside of rules.
	Allow depending on a phony that has no rule.
	Allow action prefix +.

	Implement --no-print-directory.
	New option --last-chance-rules.
	Option --no-warn centrally swallows every warn().
	Keep option handlers after 1st getopts, any might come recursively.
	Accept all unimplemented POSIX/gmake options silently.

2010-07-16  Daniel Pfeiffer  <occitan@esperanto.org>

	* Mpp/File.pm (STAT_UID, is_executable): Remove unused functions.
	(lstat_array): Skip STAT_UID.
	(read_directory): Keep the stats of files that were there
	before. (Reported by Clint O.)

	* install.pl: Start move to new pod2html ids and some fixes.

	* pod/*.pod: Complete documentation of all options.  Generate more
	readable index using new pod2html ids for all pods where it makes
	sence.

	* makepp: Complete --help generated from pod/makepp_command.pod.

2010-06-01  Daniel Pfeiffer  <occitan@esperanto.org>

	* Mpp/Signature/c_compilation_md5.pm (md5sum_c_tokens): Find
	terminating ' to avoid missed rebuilds.

	* Mpp/Rule.pm (execute_command): fork/exec on Cygwin and MinGW
	like on Unix.

	* Mpp/CommandParser/Gcc.pm (xparse_command):
	* Mpp/Scanner.pm: Eliminate useless add_dependency.

	* Mpp/Text.pm (pattern_substitution): Optimize.

	* pod/makepp_compatibility.pod: Add 5.12.1 and more CPAN-tester
	results.

2010-04-22  Daniel Pfeiffer  <occitan@esperanto.org>

Snapshot makepp-1.50.100422.tgz, version 1.50-04221-02242-02094 on CPAN.

	* Mpp/Utils.pm (Mpp::Rewrite::cwd): Work around a 5.12.0 crash.

	* install.pl:
	* pod/makepp_compatibility.pod: Mention 5.12 as working.

	* pod/makepp_command.pod: Mention --sandbox as a workaround for -j
	on native Win.

2010-02-25  Daniel Pfeiffer  <occitan@esperanto.org>

	* Mpp/Makefile.pm (parse_rule):
	* t/rule_include.test: Use :scanner none if :include file found.

	* makepp (parse_command_line): Order options alphabetically.

	* makepplog (instdir): Erstwhile workaround for 5.11 deprecation.

	* pod/makepp_rules.pod:
	* pod/makepp_faq.pod: Update about new behaviour of :include.

	* pod/makepp_repositories.pod: Remove long outdated limitation of
	only parsing the 1st command of an action.  Explain about
	--dont-read.

	* pod/makepp_compatibility.pod: Add more tested versions.

2010-02-09  Daniel Pfeiffer  <occitan@esperanto.org>

	* Mpp/Makefile.pm (parse_rule):
	* makepp (build_target_done):
	* Mpp/FileOpt.pm (set_additional_dependencies): Handle :include of
	generated dependency file. (Suggested by Yoni Londner)
	(get_rule): Remove unneeded label colliding with a reserved word
	in perl 5.11.

	* Mpp/Subs.pm (scanner_none):
	* Mpp/ActionParser.pm (parse_rule): Keep promise that scanner none
	shuts up.

	* Mpp/Rule.pm (sorted_dependencies): Use only NAME for same
	dependency given multiply.

	* t/rule_include.test: New test.

	* install.pl (highlight_keywords): Handle :include & :last_chance.

2009-12-27  Daniel Pfeiffer  <occitan@esperanto.org>

	* Mpp/Subs.pm (scanner_skip_word): Make it work per rule.
	(s_register_scanner): Make prefix "scanner_" optional and allow
	"-".

	* Mpp/Makefile.pm (parse_rule): Allow "-" in :scanner.

	* Mpp/ActionParser.pm (parse_rule): Suppress "action scanner not
	found" warning with skip-word.  Clearer suggestion.

	* Mpp/Rule.pm (parser): Simplify.


makepp/Mpp.pm
# $Id: Mpp.pm,v 1.9 2010/09/13 21:02:52 pfeiffer Exp $

=head1 NAME

Mpp - Common subs for makepp and makeppreplay

=head1 DESCRIPTION

This package contains basic stuff for makepp.

=cut

package Mpp;

use strict;

BEGIN {
  eval "sub ARCHITECTURE() { '$Config{archname}' }"; # Get a tag for the architecture.
}

use Mpp::Text;

#
# Signal handling and exiting
#
# Do this early, because the END block defined below shall be the first seen
# by perl, such that it is the last executed.  Unless we need to propagate a
# signal, it leaves the process via POSIX::_exit, so that no expensive garbage
# collection of Mpp::File objects occurs.  All other places can use die or normal
# exit.  If you define additional END blocks in any module, you must take care
# to not reset $?.
#
{
  my (%signo, @signame);
  if(defined(my $sig_name = $Config{sig_name})) {
    my $i=0;
    for my $name (split(' ', $sig_name)) {
      $signo{$name} ||= $i;
      $signame[$i] = $name;
      $i++;
    }
  }
  sub signo {
    $signo{$_[0]} || $_[0];
  }
  sub signame {
    $signame[$_[0]] || $_[0];
  }
}

our $int_signo = signo 'INT';
our $quit_signo = signo 'QUIT';

my $logfh;
our @close_fhs = \(*STDOUT, *STDERR);
my $warn_level = 1;		# Default to warning.

our $critical_sections = 0;

our $n_files_changed = 0;	# Keep track of the number of files that
				# we actually changed.
our $n_phony_targets_built = 0;	# Keep track of the number of phony targets
                                # built, too.
our $error_found = 0;		# Non-zero if we found an error.  This is used
				# to stop cleanly when we are doing a parallel
				# build (unless -k is specified).
our $failed_count = 0;		# How many targets failed.  TODO: my when moving build et al here
our $build_cache_hits = 0;	# Number of the files changed that were
				# imported from a build cache.
our $rep_hits = 0;		# Number of the files changed that were
				# imported from a rep.
our $print_directory = 1;	# Default to printing it.
our $log_level = 2;		# Default to logging. 1: STDOUT, 2: $logfile
sub log($@);

my @signals_to_handle = is_windows > 0 && is_perl_5_6 ?
  () :				# ActiveState 5.6 doesn't define signals.
  qw/INT QUIT HUP TERM/;
{
  my %pending_signals;
  sub suicide {
    my ($sig) = @_;
    $SIG{$sig} = 'DEFAULT';

    # If we're propagating a signal from a subprocess that produced a core
    # dump, then we want to propagate the same signal without overwriting
    # the core file.  This is true even if the subprocess didn't produce a
    # core dump, because it could be propagating a signal from its child
    # that did produce a core dump.
    $sig = 'INT' if $sig eq 'QUIT' &&
      !(require BSD::Resource && BSD::Resource::setrlimit( &BSD::Resource::RLIMIT_CORE, 0, 0 ));

    kill $sig, $$;
    POSIX::_exit(0x80 | signo($sig)); # just in case;
  }
  sub handle_signal {
    my $sig = $_[0];

    # Do nothing on SIGINT or SIGQUIT if there are external processes.  These
    # signals are sent to all child processes as well, and if any of those
    # processes propagates them, then we will too.  Otherwise we ignore them
    # completely (which is the UNIX convention).
    return if $Mpp::Event::n_external_processes &&
      ($sig eq 'INT' || $sig eq 'QUIT');

    $pending_signals{$sig} = 1;
    # If there's nothing that we absolutely have to do before terminating, then
    # just terminate.
    exit unless $critical_sections;

    &Mpp::Event::Process::terminate_all_processes;
  }
  # This gets called after a critical section is completed:
  sub propagate_pending_signals {
    die if $critical_sections<0;
    return if $critical_sections;
    exit 2 if grep $_, values %pending_signals;
  }
  sub reset_signal_handlers {
    @SIG{@signals_to_handle} = ( 'DEFAULT' ) x @signals_to_handle;
  }
  # Autovivification of hash elements probably isn't reentrant, so initialize
  # them here in case we're using Perl 5.6.
  @pending_signals{@signals_to_handle} = ( 0 ) x @signals_to_handle;

  END {
    Mpp::log N_REP_HITS => $rep_hits
      if $log_level && $rep_hits;
    Mpp::log N_CACHE_HITS => $build_cache_hits
      if $log_level && $build_cache_hits;
    Mpp::log N_FILES => $n_files_changed, $n_phony_targets_built, $failed_count
      if defined $logfh;		    # Don't create log for --help or --version.
    if( exists $Devel::DProf::{VERSION} ) { # Running with profiler?
      warn "Doing slow exit, needed for profiler.\n";
    } else {
      close $_ for @close_fhs;
      $pending_signals{$_} && suicide $_ for keys %pending_signals;
      POSIX::_exit $?;
    }
  }

}
@SIG{@signals_to_handle} = ( \&handle_signal ) x @signals_to_handle;

my $invocation = join_with_protection($0, @ARGV);

our $progname;

for( qw(/usr/xpg4/bin/sh /sbin/xpg4/sh /bin/sh) ) {
  if( -x ) {
    $ENV{SHELL} = $_;		# Always use a hopefully Posix shell.
    last;
  }
}
delete $ENV{PWD};		# This is dangerous.

our $indent_level = 0;		# Indentation level in the log output.
our $keep_going = 0;		# -k specified.
my $logfile;			# Other than default log file.
our $parallel_make = 0;		# True if we're in parallel make mode.
our $profile = 0;		# Log messages about execution times.
our $verbose;
our $quiet_flag;		# Default to printing informational messages.

$SIG{__WARN__} = sub {
  my $level = ($_[0] =~ /^(?:error()|info): /s) ? '' : 'warning: ';
  my $error = defined $1;
  if( $log_level == 2 ) {
    &Mpp::log() unless defined $logfh; # Only open the file.
    print $logfh "*** $level$_[0]";
  }
  print STDERR "$progname: $level$_[0]" if $error or $warn_level;
};



=head2 Mpp::log

  Mpp::log KEY => {object | array of objects | string} ...
    if $log_level;

The list of available KEYs is present in makepplog.  If you pass an non-key
str if will later be output verbatim.  The objects must have a method `name'.

This log overrides logarithm (which is not needed by makepp).  Because of
this, and because it is not exported, it must always be invoked as Mpp::log.

The log format contains a few control chars, denoted here as ^A, ^B ...

The first line is special, the invocation preceded by "logversion^A" as
explained at @obsolete_msg in makepplog.

A leading ^B is stripped, but tells makepplog to outdent, and a leading ^C to
indent.  After that a line with no ^A is considered plain text.  Else it is
split on the ^A`s.  There must be a ^A at the line end, which allows having
multine fields between ^A`s.  If the resulting fields contain ^B`s they are
lists of simple fields, else just one simple field.

The first field is a message I<key>.  The others work as follows.  If the
simple fields contain ^C`s they are ref definitions of up to 4 fields:

    ref[^Cname[^Cdir-ref[^Cdir-name]]]

The I<refs> are numbers (hex on HP/UX with 64bit pointers) and the I<names>
are to be remembered by makepplog for these numbers.  If a I<dir-name> is
present, that is remembered for I<dir-ref>, else it has been given earlier.
If a I<dir-ref> is given, that is prependended to I<name> with a slash and
remembered for I<ref>.  Else if only I<name> is given that is remembered for
I<ref> as is.  If I<ref> is alone, it has been given earlier.  Makepplog will
output the remembered name for refs between quotes.

The fields may also be plain strings if they contain no known I<ref>.  Due to
the required terminator, the strings may contain newlines, which will get
displayed as \n.  For keys that start with N_, all fields are treated as plain
numbers, even if they happen to coincide with a I<ref>.

=cut

use Mpp::File;

my $last_indent_level = 0;
sub log($@) {

  # Open the log file if we haven't yet.  Must do it dynamically, because
  # there can be messages (e.g. from -R) before handling all options.
  unless( defined $logfh ) {
    if( $log_level == 1 ) {
      (my $mppl = $0) =~ s/\w+$/makepplog/;
      -f $mppl or
	substr $mppl, 0, 0, absolute_filename( $Mpp::original_cwd ) . '/';
      open $logfh, '|' . PERL . " $mppl -pl-" or # Pass the mesages to makepplog for formatting.
	die "$progname: can't pipe to `makepplog' for verbose option--$!\n";
    } else {
      if( !$logfile ) {
	mkdir '.makepp';	# Just in case
	$logfile = '.makepp/log';
      }
      unless( open $logfh, '>', $logfile ) {
	warn "$progname: can't create log file ./$logfile--$!\n";
	$log_level = 0;
	return;
      }
    }
    push @close_fhs, $logfh;
    printf $logfh "2\01%s\nVERSION\01%s\01%vd\01%s\01\n", $invocation, $Mpp::VERSION, $^V, ARCHITECTURE;

    # If we're running with --traditional-recursive-make, then print the directory
    # when we're entering and exiting the program, because we may be running as
    # a make subprocess.

    Mpp::Rule::print_build_cwd( $CWD_INFO )
      if defined $Mpp::Recursive::traditional;

    return unless @_;		# From __WARN__
  }

  print $logfh
    join "\01",
      $indent_level == $last_indent_level ? shift() : # Cheaper than passing a slice to map
	($indent_level < $last_indent_level ? "\02" : "\03") . shift,
      map( {
	if( !ref ) {
	  $_;
	} elsif( 'ARRAY' eq ref ) {
	  join "\02", map {		# Array shall only contain objects.
	    if( exists $_->{LOGGED} ) {	# Already defined
	      int;			# The cheapest external representation of a ref.
	    } elsif( !exists $_->{'..'} ) { # not a Mpp::File
	      # TODO: These two liness are a reminder for when we store RULE_SOURCE per ref.
	      #undef $_->{LOGGED};
	      #int() . "\03" . $_->name;
	      $_->name;
	    } elsif( exists $_->{'..'}{LOGGED} ) { # Dir already defined
	      undef $_->{LOGGED};
	      int() . "\03$_->{NAME}\03" . int $_->{'..'};
	    } else {
	      undef $_->{LOGGED};
	      undef $_->{'..'}{LOGGED};
	      int() . "\03$_->{NAME}\03" .
		int( $_->{'..'} ) . "\03" . ($_->{'..'}{FULLNAME} || absolute_filename $_->{'..'});
	    }
	  } @$_;
# The rest is a verbatim copy of the map block above.  This function is heavy
# duty, and repeating code is 6% faster than calling it as a function, even
# with &reuse_stack semantics.
	} elsif( exists $_->{LOGGED} ) {	# Already defined
	  int;			# The cheapest external representation of a ref.
	} elsif( !exists $_->{'..'} ) { # not a Mpp::File
	  # TODO: These two liness are a reminder for when we store RULE_SOURCE per ref.
	  #undef $_->{LOGGED};
	  #int() . "\03" . $_->name;
	  $_->name;
	} elsif( exists $_->{'..'}{LOGGED} ) { # Dir already defined
	  undef $_->{LOGGED};
	  int() . "\03$_->{NAME}\03" . int $_->{'..'};
	} else {
	  undef $_->{LOGGED};
	  undef $_->{'..'}{LOGGED};
	  int() . "\03$_->{NAME}\03" .
	    int( $_->{'..'} ) . "\03" . ($_->{'..'}{FULLNAME} || absolute_filename $_->{'..'});
	}
      } @_ ),
      "\n";
  $last_indent_level = $indent_level;
}

=head2 flush_log

Flush the log file and standard file handles.  This is useful for making sure
that output of a perl action is not lost before the action's process
terminates with POSIX::_exit.

=cut

sub flush_log {
  my $fh = select STDOUT; local $| = 1;
  select STDERR; $| = 1;
  if( defined $logfh ) { select $logfh; $| = 1 }
  select $fh;
}

my $hires_time;
sub print_profile {
  print $profile ? "$progname: Begin \@" . &$hires_time() . " $_[0]\n" : "$_[0]\n";
}

sub print_profile_end {
  print "$progname: End \@" . &$hires_time() . " $_[0]\n";
}

=head2 print_error

  print_error "message", ...;

Prints an error message, with the program name prefixed.  For any arg which is
a reference, $arg->name is printed.

If any filename is printed, it should have be quoted as in `filename' or
`filename:lineno:' at BOL, so that IDEs can parse it.

=cut

sub print_error {
  my( $log, $stderr ) = $log_level && '*** ';
  my $str = '';
  if( $_[0] =~ /^error()/ || $_[0] !~ /^\S+?:/ ) { # No name?
    $stderr = "$progname: ";
    $str = 'error: ' unless defined $1;
  }
  $str .= ref() ? $_->name : $_
    for @_;
  $str .= "\n" if $str !~ /\n\z/;
  print STDERR $stderr ? $stderr . $str : $str;
  if( $log_level == 2 ) {
    &Mpp::log() unless defined $logfh; # Only open the file.
    print $logfh $log . $str;
  }
  &flush_log;
}


sub perform(@) {
  #my @handles = @_;		# Arguments passed to wait_for.
  my $status;
  my $start_pid = $$;
  my $error_message = $@ || '';
  unless( $error_message ) {
    eval { $status = &wait_for }; # Wait for args to be built.
				# Wait for all the children to complete.
    $error_message .= $@ if $@;	# Record any new error messages.
  }
  {
    my $orig = '';
    if($error_message) {
        chomp($orig = $error_message);
        $orig = " (Original error was $orig)";
    }

    if( $$ != $start_pid ) {
      print STDERR qq{makepp internal error: sub-process died or returned to outer scope.
If we had not caught this, it would cause exit blocks to run multiple times.
Use POSIX::_exit instead.$orig Stopped};
      close $_ for @close_fhs;
      POSIX::_exit 3;
    }

    # Wait for our last jobs to finish.  This can be the case when running
    # with -kj<n>, a job failed and we ran out of queued jobs.
    &Mpp::Event::event_loop
      while $Mpp::Event::n_external_processes;

    die qq{makepp internal error: dangling critical section.
This means that there was a live process in the background when makepp
died, so makepp did not have a chance to create build info files for
targets generated by that process.  It also means that makepp can't
propagate signals.  This could instead mean that you need an extra 'eval'
somewhere to prevent an exception from bypassing process accounting and
signal propagation.$orig Stopped}
      if $critical_sections;

    if( $error_found && $error_found =~ /^signal ($int_signo|$quit_signo)$/ ) {
      my $sig = $1;
      handle_signal signame($sig);
    }
  }

  if( $quiet_flag ) {
    # Suppress following chatter.
  } elsif( $n_files_changed || $rep_hits || $build_cache_hits || $n_phony_targets_built || $failed_count ) {
    print "$progname: $n_files_changed file" . ($n_files_changed == 1 ? '' : 's') . ' updated' .
      ($rep_hits ? ", $rep_hits repository import" . ($rep_hits == 1 ? '' : 's') : '') .
      ($build_cache_hits ? ", $build_cache_hits build cache import" . ($build_cache_hits == 1 ? '' : 's') : '') .
      ($error_found ? ',' : ' and') .
      " $n_phony_targets_built phony target" . ($n_phony_targets_built == 1 ? '' : 's') . ' built' .
      ($failed_count ? " and $failed_count target" . ($failed_count == 1 ? '' : 's') . " failed\n" : "\n");
  } elsif( !$error_message ) {
    print "$progname: no update necessary\n";
  }

  print "$progname: Ending \@" . &$hires_time() . "\n" if $profile;
  print_error $error_message if $error_message;
  exit 1 if $error_message || $status || !MAKEPP && $failed_count;
				# 2004_12_06_scancache has a use case for not failing despite $failed_count
  exit 0;
}

our @common_opts =
  (
    ['k', qr/keep[-_]?going/, \$keep_going],

    [undef, qr/log(?:[-_]?file)?/, \$logfile, 1],

    ['n', qr/(?:just[-_]?print|dry[-_]?run|recon)/, \$Mpp::dry_run],
    [undef, qr/no[-_]?log/, \$log_level, undef, 0], # Turn off logging.
    [undef, qr/no[-_]?print[-_]?directory/, \$print_directory, 0, undef],
    [undef, qr/no[-_]?warn/, \$warn_level, undef, 0],

    [undef, 'profile', \$profile, undef, sub {
       if( !$hires_time ) {
	 eval { require Time::HiRes };
	 $hires_time = $@ ? sub { time } : \&Time::HiRes::time;
	 print "$progname: Beginning \@" . &$hires_time() . "\n";
       }
     }],

    ['s', qr/quiet|silent/, \$quiet_flag],

    [qw(v verbose), undef, undef, sub {
       $verbose = 2;		# Value 2 queried only by Mpp/Cmds.
       $log_level = 1;		# Send the log to stdout instead.  Don't make
				# this the option variable, as it must be
				# exactly 1, not just true.
       $warn_level = 0;		# Warnings will be output via logging.
     }],

    splice @Mpp::Text::common_opts
  );

1;
